File: src/main/java/pl/laskowski/marcin/type/SudokuVariant.java
package pl.laskowski.marcin.type;

import pl.laskowski.marcin.model.Region;
import pl.laskowski.marcin.model.Sudoku;
import pl.laskowski.marcin.model.dividers.SudokuDivider;

import java.util.Set;

/**
 * Created by Marcin Laskowski.
 */

public abstract class SudokuVariant {

    private final int width;
    private final int height;

    public SudokuVariant(int width, int height) {
        this.width = width;
        this.height = height;
    }

    public abstract int regionSize();

    public int width() {
        return width;
    }

    public int height() {
        return height;
    }

    protected abstract SudokuDivider divider();

    public Set<Region> divideIntoRegions(Sudoku sudoku) {
        return divider().divideIntoRegions(sudoku);
    }

    public abstract Sudoku template();

}

File: app/src/main/java/com/github/ilikeyourhat/sudokusolver/model/type/SudokuVariant.kt
package com.github.ilikeyourhat.sudokusolver.model.type

import com.github.ilikeyourhat.sudokusolver.model.dividers.RegionDivider

interface SudokuType {
    val sizeX: Int
    val sizeY: Int
    val regionSize: Int
    fun divider(): RegionDivider
    fun template(): String
}